lib/ts/recipe/session/recipeImplementation.ts

```diff
@@ -162,38 +187,52 @@ export default function getRecipeInterface(
+                });
+            }
+
+            const sessionOptional = options?.sessionRequired === false;
+            logDebugMessage("getSession: optional validation: " + sessionOptional);
+
+            const preferredTransferMethod = config.getTokenTransferMethod({ req, userContext });
+
+            let transferMethod = preferredTransferMethod === "missing_auth_header" ? "header" : preferredTransferMethod;
+
+            let accessToken = getToken(req, "access", transferMethod);
             if (accessToken === undefined) {
-                // We are checking if we could've gone ahead with validation if the transferMethod was different
-                // However, we don't want to do the fallback here, but force a call to refresh
-                // This is done to ensure that the browsers update the transfermethod in a timely manner (basically on the next API call)
-                // instead of waiting for the session to expire
-                accessToken = getToken(
-                    config,
-                    req,
-                    "access",
-                    userContext,
-                    transferMethod === "cookie" ? "header" : "cookie"
-                );
+                const fallbackMethod = transferMethod === "cookie" ? "header" : "cookie";
+                // We are checking if we could go ahead with validation if the transferMethod was different
+                // However, we don't want to allow fallback here if there is an up-to-date frontend SDK that sends the auth-mode header.
+                accessToken = getToken(req, "access", fallbackMethod);
                 if (accessToken !== undefined) {
-                    logDebugMessage(
-                        "getSession: Returning try refresh token because the access token was not sent as a " +
-                            transferMethod
-                    );
-                    throw new STError({
-                        message: "access token sent with the wrong method. Please call the refresh API",
-                        type: STError.TRY_REFRESH_TOKEN,
-                    });
+                    if (preferredTransferMethod === "missing_auth_header") {
+                        transferMethod = fallbackMethod;
+                        logDebugMessage(
+                            `getSession: falling back to use ${fallbackMethod} as an authentication-method`
+                        );
+                    } else {
+                        logDebugMessage(
+                            `getSession: returning TRY_REFRESH_TOKEN because preferred auth-mode doesn't allow fallback and access token was sent with the wrong method`
+                        );
+                        throw new STError({
+                            message: "access token sent with the wrong method. Please call the refresh API",
+                            type: STError.TRY_REFRESH_TOKEN,
+                        });
+                    }
                 }
```

Comment:
The code flow for this block is a little strange:
- If the `preferredTransferMethod !== missing_auth_header`, then we should actually keep accessToken as undefined so that the code below runs and the user either gets undefined or unauthorised error based on sessionoptional.
- If the `preferredTransferMethod !== missing_auth_header`, then we shouldn't even bother with querying `getToken(req, "access", fallbackMethod);`? So ideally the `if` condition should be `if (accessToken === undefined && preferredTransferMethod === "missing_auth_header") {`?